src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp

```diff
@@ -0,0 +1,173 @@
+/**
+ * @file bilinear_interpolation_impl.hpp
+ * @author Kris Singh and Shikhar Jaiswal
+ *
+ * Implementation of the bilinear interpolation function as an individual layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_BILINEAR_INTERPOLATION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BILINEAR_INTERPOLATION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "bilinear_interpolation.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+template<typename InputDataType, typename OutputDataType>
+BilinearInterpolation<InputDataType, OutputDataType>::
+BilinearInterpolation():
+    inRowSize(0),
+    inColSize(0),
+    outRowSize(0),
+    outColSize(0),
+    depth(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+BilinearInterpolation<InputDataType, OutputDataType>::
+BilinearInterpolation(
+    const size_t inRowSize,
+    const size_t inColSize,
+    const size_t outRowSize,
+    const size_t outColSize,
+    const size_t depth):
+    inRowSize(inRowSize),
+    inColSize(inColSize),
+    outRowSize(outRowSize),
+    outColSize(outColSize),
+    depth(depth)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BilinearInterpolation<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, 1);
+
+  assert(input.n_rows == inRowSize * inColSize * depth);
+  assert(input.n_cols == 1);
+  assert(output.n_rows == outRowSize * outColSize * depth);
+  assert(output.n_cols == 1);
+  assert(inRowSize >= 2);
+  assert(inColSize >= 2);
+
+  arma::cube inputAsCube(input.memptr(), inRowSize, inColSize, depth);
+  arma::cube outputAsCube(output.memptr(), outRowSize, outColSize, depth, false, true);
+
+  double scaleRow = (double) inRowSize / (double) outRowSize;
+  double scaleCol = (double) inColSize / (double) outColSize;
+
+  arma::mat22 coeffs;
+  for (size_t i = 0; i < outRowSize; i++)
+  {
+    for (size_t j = 0; j < outColSize; j++)
+    {
+      size_t rOrigin = (size_t) std::floor(i * scaleRow);
+      size_t cOrigin = (size_t) std::floor(j * scaleCol);
+
+      if (rOrigin > inRowSize - 2)
+        rOrigin = inRowSize - 2;
+      if (cOrigin > inColSize - 2)
+        cOrigin = inColSize - 2;
+
+      double deltaR = i * scaleRow - rOrigin;
+      double deltaC = j * scaleCol - cOrigin;
+      coeffs[0] = (1 - deltaR) * (1 - deltaC);
```

Comment:
Not sure this is acutely faster if we also move this one loop up and just multiply by ``(1 - deltaC)`` here.

Reply:
`(1 - deltaC)` would change at every iteration of `j`. Not sure if this would be a good idea, as I'd also have to divide by the same at the end of the each iteration of `j`. Only other way I can think of breaking this up is to store the `deltaR`'s and `deltaC`'s in different vectors and letting element wise multiplication do the job here, but that won't be faster than this.

Reply:
Right my idea was, to set ``coeffs[0] = (1 - deltaR)`` inside the first loop and ``do coeffs[0] *= (1 - deltaC);`` here. Maybe I missed something?

Reply:
`deltaC` is defined as:
```
double deltaC = j * scaleCol - cOrigin;
```
which changes at every `j = 0, 1, 2...`. Now doing `coeffs[0] *= (1 - deltaC)` inside every `j`th iteration, would accumulate all the `(1 - deltaC)` values over the iterations, which can only be corrected if at the end of every `j`th iteration, I do `coeffs[0] /= (1 - deltaC)`, to get back `coeffs[0] = (1 - deltaR)`, as in the first loop.

Reply:
Ahh, your are right, thanks for the clarification.